### PR TITLE
Fix to correctly support the JOIN stages for field lineage.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-13888
Encoded stage name in the input fields of operations so that we can traverse back the origin correctly.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java [hunk 1]
```diff
@@ -173,9 +175,32 @@ private void addMergeOperation(Set<String> stageInputs,
...
     for (String field : fields) {
+      if (noMergeRequiredStages.contains(currentStage)) {
+        // Current stage is of type JOIN.
+        // JOIN creates operation with input fields in format <stagename.fieldname>
+        // if the field is directly read from the schema of the one of the input stage.
+        // However it is also possible to have operation with input not tagged with the stage name, for example
+        // in case if rename is done on the join keys. In this case Join operation will read fields of the form
+        // <stagename.fieldname> however rename operations followed by it will simply read the output of the
+        // Join which wont be tagged with stagename.
+        Iterable<String> stageFieldPair = Splitter.on(".").omitEmptyStrings().trimResults().split(field);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java]: Stage names and field names are both allowed to have '.' in them. There is no way to do this correctly with a separator. Is there a jira open to do join lineage properly?  If so, can you add a TODO with the jira number. If not, can you open a jira and add a TODO with the number.

A slightly better, but still bad option is to use an uncommon character as a separator.

Another slightly better option is to check for the entire stage name plus a '.' as a prefix, and use that to split into stage name and field name.

Either way there are still scenarios where things will be buggy, such as when a field name happens to match the stage name.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java [hunk 2]
```diff
@@ -112,7 +112,7 @@ public DefaultFieldLineageReader(DatasetFramework datasetFramework, TransactionS
         : fieldLineageDataset.getOutgoingOperations(endPointField.getEndPoint(), start, end);
     });
 
-    Set<ProgramRunOperations> endPointFieldOperations = new HashSet<>();
+    Set<ProgramRunOperations> endPointFieldOperations = new LinkedHashSet<>();
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java]: why is this needed

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java [hunk 3]
```diff
@@ -319,7 +320,7 @@ private void addOperationReferenceRecord(byte[] direction, EndPoint endPoint, Pr
     Map<Long, Set<ProgramRunId>> checksumsWithProgramRunsInRange
       = getChecksumsWithProgramRunsInRange(direction, endPoint, start, end);
 
-    Set<ProgramRunOperations> result = new HashSet<>();
+    Set<ProgramRunOperations> result = new LinkedHashSet<>();
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:323]: why is this needed

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java [hunk 4]
```diff
@@ -347,7 +348,7 @@ private void addOperationReferenceRecord(byte[] direction, EndPoint endPoint, Pr
     byte[] scanStartKey = getScanKey(direction, endPoint, end);
     // time is inverted, hence we need to pass start-time for getting end key
     byte[] scanEndKey = getScanKey(direction, endPoint, start);
-    Map<Long, Set<ProgramRunId>> result = new HashMap<>();
+    Map<Long, Set<ProgramRunId>> result = new LinkedHashMap<>();
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:351]: why is this needed

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 5]
```diff
@@ -585,21 +586,40 @@ public void destroy() {
...
+                                     .collect(Collectors.toSet()));
+          }
         }
       }
 
       Schema outputSchema = stageSpec.getOutputSchema();
       if (outputSchema != null && outputSchema.getFields() != null) {
-        builder.addStageOutputs(outputSchema.getFields().stream().map(Schema.Field::getName)
+        stageOutputs.addAll(outputSchema.getFields().stream().map(Schema.Field::getName)
                                   .collect(Collectors.toSet()));
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java (hunk 1 above)
Comment (sagarkapare) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java]: I see. Added more stricter check that prefix is a part of stage name per suggestion. Also added a TODO (Join API).

On a separate note, I implemented this logic based on the current `Joiner` plugin which asks user to specify join keys and aliases as `StageName.fieldname`. So if either of stage name or field name has `.` in it then the plugin won't work right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java (hunk 2 above)
Comment (sagarkapare) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java]: Added this to maintain the ordering within the different set of operations. This allows the returned set of operations by REST api to be sorted by time descending.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java]: If the Joiner is using '.' to split, it is also wrong. It should be using the input stages as the prefixes. Please open a jira for the joiner if it is also doing the wrong thing.

However, in the joiner, it is unambiguous because it expects everything will be prefixed by stage name.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java (hunk 3 above)
Comment (sagarkapare) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:323]: same. To preserve the descending ordering.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java (hunk 4 above)
Comment (sagarkapare) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java:351]: ditto.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java (hunk 2 above)
Comment (albertshau) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java]: is this just for convenience, or is this part of the contract. For example, does the UI rely on this behavior? If it is part of the contract, it should be documented in the javadocs.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java [hunk 6]
```diff
@@ -189,9 +189,12 @@ private void addMergeOperation(Set<String> stageInputs,
...
         // <stagename.fieldname> however rename operations followed by it will simply read the output of the
         // Join which wont be tagged with stagename.
+
+        //TODO CDAP-13298 Field names and stage names can have '.' in them.
+        // Probably better to support new JOIN operation type which takes stage name and field name.
         Iterable<String> stageFieldPair = Splitter.on(".").omitEmptyStrings().trimResults().split(field);
-        if (Iterables.size(stageFieldPair) > 1) {
-          String stageName = Iterables.get(stageFieldPair, 0);
+        String stageName = Iterables.get(stageFieldPair, 0);
+        if (Iterables.size(stageFieldPair) > 1 && stageOperations.keySet().contains(stageName)) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java]: this check on the keyset doesn't help because the field has already been split on '.' I checked the Joiner plugin and it looks like the Joiner plugin has the same issue like you mentioned. Let's mark this as TODO and fix it in 5.0.1, seeing how the only plugin we have has the same problem.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java [hunk 7]
```diff
@@ -173,9 +175,35 @@ private void addMergeOperation(Set<String> stageInputs,
...
+        // if the field is directly read from the schema of the one of the input stage.
+        // However it is also possible to have operation with input not tagged with the stage name, for example
+        // in case if rename is done on the join keys. In this case Join operation will read fields of the form
+        // <stagename.fieldname> however rename operations followed by it will simply read the output of the
+        // Join which wont be tagged with stagename.
+
+        //TODO CDAP-13298 Field names and stage names can have '.' in them.
+        // Probably better to support new JOIN operation type which takes stage name and field name.
+        Iterable<String> stageFieldPair = Splitter.on(".").omitEmptyStrings().trimResults().split(field);
+        String stageName = Iterables.get(stageFieldPair, 0);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java]: I'm not a big fan of the Iterables class. It's usage here can be replaced by getting the iterator, then calling next() on this line, hasNext() instead of Iterables.size(), then next() instead of get(..., 1).

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-RUT1588

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java (hunk 7 above)
Comment (sagarkapare) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java]: fixed.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java (hunk 6 above)
Comment (sagarkapare) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsProcessor.java]: Added TODO on line 193.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java (hunk 2 above)
Comment (sagarkapare) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java]: UI does not rely on this, since it displays whatever operations are returned by backend which is undefined. However its better to return them in the order. Added javadocs.

**Comment by sagarkapare** [comment]
@albertshau addressed comments. please look again.thanks!

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by sagarkapare** [comment]
Build is green here - https://builds.cask.co/browse/CDAP-RUT1588-6. Squashing and merging.
